src/main/java/io/reactivex/internal/operators/flowable/FlowableSwitchMapCompletable.java

```diff
@@ -0,0 +1,239 @@
+ */
+
+package io.reactivex.internal.operators.flowable;
+
+import java.util.concurrent.atomic.AtomicReference;
+
+import org.reactivestreams.Subscription;
+
+import io.reactivex.*;
+import io.reactivex.annotations.Experimental;
+import io.reactivex.disposables.Disposable;
+import io.reactivex.exceptions.Exceptions;
+import io.reactivex.functions.Function;
+import io.reactivex.internal.disposables.DisposableHelper;
+import io.reactivex.internal.functions.ObjectHelper;
+import io.reactivex.internal.subscriptions.SubscriptionHelper;
+import io.reactivex.internal.util.*;
+import io.reactivex.plugins.RxJavaPlugins;
+
+/**
+ * Maps the upstream values into {@link CompletableSource}s, subscribes to the newer one while
+ * disposing the subscription to the previous {@code CompletableSource}, thus keeping at most one
+ * active {@code CompletableSource} running.
+ *
+ * @param <T> the upstream value type
+ * @since 2.1.11 - experimental
+ */
+@Experimental
+public final class FlowableSwitchMapCompletable<T> extends Completable {
+
+    final Flowable<T> source;
+
+    final Function<? super T, ? extends CompletableSource> mapper;
+
+    final boolean delayErrors;
+
+    public FlowableSwitchMapCompletable(Flowable<T> source,
+            Function<? super T, ? extends CompletableSource> mapper, boolean delayErrors) {
+        this.source = source;
+        this.mapper = mapper;
+        this.delayErrors = delayErrors;
+    }
+
+    @Override
+    protected void subscribeActual(CompletableObserver s) {
+        source.subscribe(new SwitchMapCompletableObserver<T>(s, mapper, delayErrors));
+    }
+
+    static final class SwitchMapCompletableObserver<T> implements FlowableSubscriber<T>, Disposable {
+
+        final CompletableObserver downstream;
+
+        final Function<? super T, ? extends CompletableSource> mapper;
+
+        final boolean delayErrors;
+
+        final AtomicThrowable errors;
+
+        final AtomicReference<SwitchMapInnerObserver> inner;
+
+        static final SwitchMapInnerObserver INNER_DISPOSED = new SwitchMapInnerObserver(null);
+
+        volatile boolean done;
+
+        Subscription upstream;
+
+        SwitchMapCompletableObserver(CompletableObserver downstream,
+                Function<? super T, ? extends CompletableSource> mapper, boolean delayErrors) {
+            this.downstream = downstream;
+            this.mapper = mapper;
+            this.delayErrors = delayErrors;
+            this.errors = new AtomicThrowable();
+            this.inner = new AtomicReference<SwitchMapInnerObserver>();
+        }
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            if (SubscriptionHelper.validate(upstream, s)) {
+                this.upstream = s;
+                downstream.onSubscribe(this);
+                s.request(Long.MAX_VALUE);
+            }
+        }
+
+        @Override
+        public void onNext(T t) {
+            CompletableSource c;
+
```

Comment:
hm, I'm looking at `Flowable.switchMap` and `Observable.switchMap` and they dispose current first:

https://github.com/ReactiveX/RxJava/blob/8068404179b0d2e07da6f0e10ea95110d98e118d/src/main/java/io/reactivex/internal/operators/flowable/FlowableSwitchMap.java#L109

https://github.com/ReactiveX/RxJava/blob/8068404179b0d2e07da6f0e10ea95110d98e118d/src/main/java/io/reactivex/internal/operators/observable/ObservableSwitchMap.java#L105

While this operator maps first and then disposes current which give current more time to work which makes it more `delayErrors` friendly I guess, however I still have a question about `delayErrors` below

Reply:
I've tried to adapt `Flowable.switchMap` behavior to respect `delayErrors` if mapper throws but failed so far, this is the test I wrote, maybe it'll help you:

```java
@Test
public void delayErrorsErrorInMapper() {
    final PublishProcessor<Object> trigger = PublishProcessor.create();

    TestSubscriber<Integer> ts = Flowable
            .just(1, 2)
            .switchMapDelayError(new Function<Integer, Publisher<Integer>>() {
                @Override
                public Publisher<Integer> apply(Integer integer) throws Exception {
                    if (integer.equals(1)) {
                        return Flowable
                                .combineLatest(Flowable.just(1), trigger, new BiFunction<Integer, Object, Integer>() {
                                    @Override
                                    public Integer apply(Integer integer, Object o) {
                                        return integer;
                                    }
                                })
                                .doOnCancel(new Action() {
                                    @Override
                                    public void run() throws Exception {
                                        trigger.toString();
                                    }
                                });
                    } else {
                        // Fail for second item.
                        throw new TestException("Item " + integer + " mapping failure.");
                    }
                }
            })
            .test();

    ts.assertNoValues();
    ts.assertNotTerminated();

    trigger.onNext(new Object());

    ts.assertValue(1);
    ts.assertError(new Predicate<Throwable>() {
        @Override
        public boolean test(Throwable throwable) {
            return throwable instanceof TestException && throwable.getMessage().equals("Item 2 mapping failure.");
        }
    });
}
```

Reply:
Yes, essentially it gives a larger window for errors but otherwise has no effect because there are no items to forward and a late completion has no effect anyway.

Mapping errors are terminal errors, as if the upstream ended with a failure. The problem is that the operator can't decide if it is okay to continue. You can always do `try-catch` and return a `Completable.error`.